vars/pythonSetupPyPipeline.groovy

```diff
@@ -9,148 +9,310 @@
+                //   will be installed and the doc generation will fail due to not found imports.
+                sh "python setup.py develop --user"
+                sh "python setup.py build_sphinx"
+              }
+            }
+          } // Build docs
         }
-      }
+      } // Build
 
-      stage("Execute Tests") {
-        steps {
-          sh "python setup.py test --addopts '--cov-report xml:build/coverage.xml --cov-report term --cov-branch --junitxml=build/test_results.xml'"
-        }
-        post {
-          always {
-            junit 'build/test_results.xml'
-            cobertura coberturaReportFile: 'build/coverage.xml'
-          }
-        }
-      }
+      stage("Validate") {
+        parallel {
+          stage("Run Linter") {
+            agent {
+              docker {
+                image buildImage.imageName()
+                args pipelineParams.dockerRunArgs
+                reuseNode true
+              }
+            }
+            steps {
+              // Necessary to avoid issues with creating pylint files that track the delta of warnings.
+              withEnv(["HOME=$WORKSPACE"]) {
+                // setuptools-lint does not do a good job in installing all its dependencies.
+                // use `develop` over `install` to avoid creating an egg file in dist/
+                sh "python setup.py develop --user"
+                sh "python setup.py lint --lint-output-format parseable"
+              }
+            }
+            post {
+              always {
+                recordIssues enabledForFailure: true, tool: pyLint(), healthy: 0
+              }
+            }
+          } // Run Linter
 
-      stage("Build Docs") {
-        steps {
-          // Necessary to be able to locally install packages, which is required for build_sphinx
-          withEnv(["HOME=$WORKSPACE"]) {
-            sh "python setup.py install --user"
-            sh "python setup.py build_sphinx"
-          }
+          stage("Run Tests") {
+            agent {
+              docker {
+                image buildImage.imageName()
+                args pipelineParams.dockerRunArgs
+                reuseNode true
+              }
+            }
+            steps {
+              sh "python setup.py test --addopts '--cov-report xml:build/coverage.xml --cov-report term --cov-branch --junitxml=build/test_results.xml'"
+            }
+            post {
+              always {
+                junit "build/test_results.xml"
+                cobertura coberturaReportFile: "build/coverage.xml"
+              }
+            }
+          } // Run Tests
         }
-      }
+      } // Validate
 
       stage("Package") {
+        agent {
+          docker {
+            image buildImage.imageName()
+            args pipelineParams.dockerRunArgs
+            reuseNode true
+          }
+        }
         steps {
           sh "python setup.py sdist"
           sh "twine check dist/*"
+
+          script {
```

Comment:
Nope, since then we would not discover issues in for example malformed markdown format of the README or other issues during a regular build. The Docker deploy stage is only executed on-demand in branches/PRs and always on master